doc/developer_guide/parsing.rst

```diff
@@ -137,6 +137,129 @@ create the appropriate `Arg` instance. Previously we relied on the
+
+.. highlight:: fortran
+
+For example::
+
+    program alg
+      use kern_mod, only kern_type
+      real(32) :: a
+      real(64) :: b
+      call invoke(kern_type(a), kern_type(b))
+    end program alg
+
+    module kern_mod
+      type kern_type
+        ! metadata description
+	procedure :: kern_code
+      end type kern_type
+      interface kern_code
+	kern_code_32
+        kern_code_64
+      end interface kern_code
+    contains
+      subroutine kern_code_32(arg)
+        real(32) :: arg
+      end subroutine kern_code_32
+      subroutine kern_code_64(arg)
+        real(64) :: arg
+      end subroutine kern_code_64
+    end module kern_mod
+
+In the above example, the first call to kern will call `kern_code_32`
+and the second will call `kern_code_64`. Note, the actual code will
+use types for arguments in the algorithm layer, but for clarity
+precision has been used.
+
+In order to support the above mixed precision kernel implementation,
+the PSy-layer generated by PSyclone must declare data passed into an
+invoke call with the same precision as is declared in algorithm layer
+i.e., in the above example, variable `a` must be declared as
+`real(32)` and variable `b` as `real(64)`. The only way for PSyclone
+to determine the required precision for a variable is by extracting
+the information from the algorithm layer.
+
+To support the extraction of precision information, the name and
+precision of a variables are stored in a dictionary when the algorithm
+layer is parsed (see `psyclone.parser.algorithm.py`). This allows
+PSyclone to look up the precision of a variable when it is specified
+in an `invoke` call. The reason for implementing support in this way
+is because `fparser2` does not currently support a symbol table,
+therefore there is no link between variables names and their types. In
+the future, the algorithm layer will be translated into PSyIR, which
+does have a symbol table, so this dictionary will no longer be
+required (see issue #753).
+
+All declarations that specify variables or types are stored in the
+dictionary, including those specified within locally defined
+types. Variables may also be single valued or arrays.
+
+.. highlight:: fortran
+
+For example::
+
+    program alg
+      real(kind=r_def) :: rscalar
+      type(my_type) :: field(10)
+      type fred
+        integer(kind=i_def) :: iscalar
+      end type fred
+    end program alg
+
+As the current implementation only stores variable names and does not
+know about variable scope there is a restriction that any variables
+within an algorithm code with the same name must have the same
+precision/type. This restriction will be removed when the algorithm
+layer is translated to PSyIR (see issue #753). The current
+implementation could be improved but in practice the lfric code does
+not fall foul of this restriction.
+
+There is also a constraint that invoke arguments can not be
+expressions or functions if they include a variable as it is then
```

Comment:
Presumably, any function is forbidden, whether or not it takes a variable as argument?

Reply:
I'm not sure what I was trying to say there. Removed.